Support native PySpark.sql on Pandera

NeerajMalhotra-QB opened the pull request:
Finally this branch as 'almost' all the changes to support PySpark.sql on Pandera. It also comes with additional functionalities as follows:

- [x] Support native PySpark.sql on Pandera
- [x] kill switch to enable/disable on production. **(only for pyspark.sql schemas)**
- [x] control the depth of validation (SCHEMA_ONLY, DATA_ONLY, SCHEMA_DATA_BOTH). **(only for pyspark.sql schemas)**
- [x] optional metadata dict object at column, field, schema and model levels for pyspark.sql and pandas schemas both
- [x] lint and mypy fixes

Super excited for this major milestone. In coming days, we will be release documentation around above changes.

cosmicBboy commented:
can confirm that the failing test only happens with pyspark > 3.2.4

looking into it...

cosmicBboy commented:
@NeerajMalhotra-QB I addressed the failing unit tests, looks like there are still some linting issues

cosmicBboy commented:
looks like there are also some docs failures: https://github.com/unionai-oss/pandera/actions/runs/5159629847/jobs/9294789864?pr=1213

FYI there are some make targets that let you build the docs locally for testing purposes: https://github.com/unionai-oss/pandera/blob/main/Makefile#L26-L31

NeerajMalhotra-QB commented:
> looks like there are also some docs failures: https://github.com/unionai-oss/pandera/actions/runs/5159629847/jobs/9294789864?pr=1213
> 
> FYI there are some make targets that let you build the docs locally for testing purposes: https://github.com/unionai-oss/pandera/blob/main/Makefile#L26-L31

sure, will take a look at it.

cosmicBboy commented:
I'll take a look.

Can you update the docs to fix the docs failures? https://github.com/unionai-oss/pandera/actions/runs/5180399308/jobs/9334552071?pr=1213. It's due to the new `metadata` property.

You can check this locally with `make docs`.

NeerajMalhotra-QB commented:
> I'll take a look.
> 
> Can you update the docs to fix the docs failures? https://github.com/unionai-oss/pandera/actions/runs/5180399308/jobs/9334552071?pr=1213. It's due to the new `metadata` property.
> 
> You can check this locally with `make docs`.

absolutely, will do.

NeerajMalhotra-QB commented:
> 
@cosmicBboy 
Its still running for others but hopefully it seems like resolved for both pyspark and pandas.

https://github.com/unionai-oss/pandera/actions/runs/5182895439/jobs/9340171409?pr=1213

cosmicBboy commented:
Note that docs CI is only run for python 3.8 and 3.9: https://github.com/unionai-oss/pandera/blob/main/.github/workflows/ci-tests.yml#L235-L241

So the failure in this CI run still needs to be fixed: https://github.com/unionai-oss/pandera/actions/runs/5182895439/jobs/9340171657?pr=1213

NeerajMalhotra-QB commented:
@cosmicBboy / Niels, any idea about this one?

https://github.com/unionai-oss/pandera/actions/runs/5191836192/jobs/9360195108

cosmicBboy commented:
[https://github.com/NeerajMalhotra-QB/pandera/blob/union/pandera/api/extensions.py#L144-L148 this change needs to be reverted ](https://github.com/NeerajMalhotra-QB/pandera/blob/union/pandera/api/extensions.py#L217-L220) needs to include `ps.DataFrame` if pyspark is installed.

We should add a test for this in the unit tests, but good thing the docs caught it!

NeerajMalhotra-QB commented:
@cosmicBboy / Niels, I think issues related to docs and doctests are resolved now: https://github.com/unionai-oss/pandera/actions/runs/5193388249/jobs/9363869457?pr=1213

I am also seeing failures on `pyspark.pandas` related tests. Can you please take a look when possible?
https://github.com/unionai-oss/pandera/actions/runs/5193388249/jobs/9363869160?pr=1213

cosmicBboy commented:
@NeerajMalhotra-QB tests are now passing!

The broken tests were actually due to issues in `test_pyspark_params.py` (which I renamed to [test_pyspark_config.py](https://github.com/unionai-oss/pandera/pull/1213/files#diff-515a6c07aa238686e9c29efbe79b9da43f747393648133fab82fc5e986f7b240))... the fact that we needed to do the whole del module cache thing made me backtrack to the way `ParamsConfig` was done... then I realized the problem of converting env vars to config is exactly the problem that [pydantic solves](https://docs.pydantic.dev/latest/usage/settings/)...

So I took the liberty of simplifying the way the config works, with a `pandera.config` module, see [here](https://github.com/unionai-oss/pandera/pull/1213/files#diff-c056bd758e1a4e542f5b3c99b58b36ace39191fb76d7186d2f7574d73ad5dbad).

I also expose a global `pandera.config.CONFIG` variable, which is an instance of `PanderaConfig` (the new parameter config class), and refactored `validate_scope` and a bunch of other places to use this global value.

I'm not sure if you wanted a per-backend configuration, which we can certainly do, but as the code was only using a global configuration this seemed like the simplest solution. Let me know your thoughts!

NeerajMalhotra-QB commented:
> @NeerajMalhotra-QB tests are now passing!
> 
> The broken tests were actually due to issues in `test_pyspark_params.py` (which I renamed to [test_pyspark_config.py](https://github.com/unionai-oss/pandera/pull/1213/files#diff-515a6c07aa238686e9c29efbe79b9da43f747393648133fab82fc5e986f7b240))... the fact that we needed to do the whole del module cache thing made me think that it shouldn't be so hard to test a piece for functionality. I then backtracked to the way `ParamsConfig` was done... then I realized the problem of converting env vars to config is exactly the problem that [pydantic solves](https://docs.pydantic.dev/latest/usage/settings/)...
> 
> So I took the liberty of simplifying the way the config works, with a `pandera.config` module, see [here](https://github.com/unionai-oss/pandera/pull/1213/files#diff-c056bd758e1a4e542f5b3c99b58b36ace39191fb76d7186d2f7574d73ad5dbad).
> 
> I also expose a global `pandera.config.CONFIG` variable, which is an instance of `PanderaConfig` (the new parameter config class), and refactored `validate_scope` and a bunch of other places to use this global value.
> 
> I'm not sure if you wanted a per-backend configuration, which we can certainly do, but as the code was only using a global configuration this seemed like the simplest solution. Let me know your thoughts!

hi @cosmicBboy, 
It looks good to me and I like the idea of pandera.config as we can leverage it for `pandas` and other backends. I don't think we need different configs for different backends. In fact it should be one config for entire setup irrespective of what client is using in pipeline e.g.`pyspark` or `pandas`. It's good that we can globalize this under pandera.config.

cosmicBboy commented:
okay! will merge this by the end of day... I will likely do another look-over to see if there are any dangling abstractions, e.g. I don't think [ColumnBase](https://github.com/NeerajMalhotra-QB/pandera/blob/union/pandera/typing/pyspark_sql.py#L54) and the [Column](https://github.com/NeerajMalhotra-QB/pandera/blob/union/pandera/typing/pyspark_sql.py#L81) generic types of needed anymore, correct?

If so, we can clean up things like that next week. If you can help identify other clean-up points that would be great too.

NeerajMalhotra-QB commented:
> 
Sure, I can take a look at it next week including ColumnBase/Column (may be in new branch). I just want to ensure removing it doesn't break anything but for now we should be good to merge this branch and next week, we can push another branch on `dev` before release cut out. agree?

cosmicBboy reviewed (approved):
🚀
